src/pal/src/arch/arm/context2.S

```diff
@@ -0,0 +1,203 @@
+#define CONTEXT_R7           CONTEXT_R6+4
+#define CONTEXT_R8           CONTEXT_R7+4
+#define CONTEXT_R9           CONTEXT_R8+4
+#define CONTEXT_R10          CONTEXT_R9+4
+#define CONTEXT_R11          CONTEXT_R10+4
+#define CONTEXT_R12          CONTEXT_R11+4
+#define CONTEXT_Sp           CONTEXT_R12+4
+#define CONTEXT_Lr           CONTEXT_Sp+4
+#define CONTEXT_Pc           CONTEXT_Lr+4
+#define CONTEXT_Cpsr         CONTEXT_Pc+4
+#define CONTEXT_Fpscr        CONTEXT_Cpsr+4
+#define CONTEXT_Padding      CONTEXT_Fpscr+4
+#define CONTEXT_D0           CONTEXT_Padding+4
+#define CONTEXT_D1           CONTEXT_D0+8
+#define CONTEXT_D2           CONTEXT_D1+8
+#define CONTEXT_D3           CONTEXT_D2+8
+#define CONTEXT_D4           CONTEXT_D3+8
+#define CONTEXT_D5           CONTEXT_D4+8
+#define CONTEXT_D6           CONTEXT_D5+8
+#define CONTEXT_D7           CONTEXT_D6+8
+#define CONTEXT_D8           CONTEXT_D7+8
+#define CONTEXT_D9           CONTEXT_D8+8
+#define CONTEXT_D10          CONTEXT_D9+8
+#define CONTEXT_D11          CONTEXT_D10+8
+#define CONTEXT_D12          CONTEXT_D11+8
+#define CONTEXT_D13          CONTEXT_D12+8
+#define CONTEXT_D14          CONTEXT_D13+8
+#define CONTEXT_D15          CONTEXT_D14+8
+#define CONTEXT_D16          CONTEXT_D15+8
+#define CONTEXT_D17          CONTEXT_D16+8
+#define CONTEXT_D18          CONTEXT_D17+8
+#define CONTEXT_D19          CONTEXT_D18+8
+#define CONTEXT_D20          CONTEXT_D19+8
+#define CONTEXT_D21          CONTEXT_D20+8
+#define CONTEXT_D22          CONTEXT_D21+8
+#define CONTEXT_D23          CONTEXT_D22+8
+#define CONTEXT_D24          CONTEXT_D23+8
+#define CONTEXT_D25          CONTEXT_D24+8
+#define CONTEXT_D26          CONTEXT_D25+8
+#define CONTEXT_D27          CONTEXT_D26+8
+#define CONTEXT_D28          CONTEXT_D27+8
+#define CONTEXT_D29          CONTEXT_D28+8
+#define CONTEXT_D30          CONTEXT_D29+8
+#define CONTEXT_D31          CONTEXT_D30+8
+
+// Incoming:
+//  r0: Context*
+//
+LEAF_ENTRY CONTEXT_CaptureContext, _TEXT
+    // Ensure we save these registers
+    push {r4-r11}
+    // Save processor flags before calling any of the following 'test' instructions
+    // because they will modify state of some flags
+    push {r1}
+    mrs r1, apsr // Get APSR - equivalent to eflags
+    push {r1} // Save APSR
+    END_PROLOGUE
+    
+    push {r2}
+    ldr r2, [R0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_INTEGER)
+    pop {r2}
+    
+    // Add 4 to stack so we point at R1, pop, then sub 8 to point at APSR
+    add sp, sp, #4
+    pop {r1}
+    sub sp, sp, #8
+    
+    strne r0, [r0, #(CONTEXT_R0)]
+    addne r0, CONTEXT_R1
+    stmiane r0, {r1-r12}
+    subne r0, CONTEXT_R1
+    
+    ldr r2, [r0, #(CONTEXT_ContextFlags)]
+    tst r2, #(CONTEXT_CONTROL)
+    
+    addne sp, sp, #(10*4) // This needs to put the stack in the same state as it started
+    strne sp, [r0, #(CONTEXT_Sp)]
+    subne sp, sp, #(10*4)
+    
+    strne lr, [r0, #(CONTEXT_Lr)]
```

Comment:
That's something I hadn't considered, not too sure how we can get the LR of the caller however because I don't think the ABI specifies how the LR must be preserved.